### PR TITLE
HLM-5169 adding tourstep

### PR DESCRIPTION
**Pull request by siddhant-nawale-egov**

### CODE REVIEW

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 1]
```diff
@@ -88,19 +93,41 @@ const Upload = ({
...
+  }, []);
+
   // UseEffect for checking completeness of data before moveing to next section
   useEffect(() => {
     if (!fileDataList || checkDataCompletion !== "true" || !setCheckDataCompletion) return;
-    const valueList = fileDataList ? Object.values(fileDataList) : [];
-    if (valueList.length !== 0 && fileDataList.Population?.error === null) setCheckDataCompletion("valid");
-    else setCheckDataCompletion("invalid");
+    if (!microplanData?.upload || !_.isEqual(fileDataList, microplanData.upload)) setModal("data-change-check");
+    else updateData();
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:110]: Can you please tell me the different between your suggestion and what i have done

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/MicroplanPreview.js [hunk 2]
```diff
@@ -119,9 +137,71 @@ const MicroplanPreview = ({
...
+      campaignId,
+      UpdateMutate,
+      setToast,
+      updateData,
+      setLoderActivation,
+      navigationEvent?.name === "next"?"GENERATED":"DRAFT",
+      t
+    );
+    setModal("none");
+  };
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/MicroplanPreview.js:195]: It has only 2 responsibilities
1) call a function that updates microplan data
2) call a api function
Should i destructure it mor

[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/MicroplanPreview.js:195]: But this is just we making a new function to just call 1 function

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 3]
```diff
@@ -120,14 +147,16 @@ const Upload = ({
...
   const handleKeyPress = (event) => {
     // if (modal !== "upload-guidelines") return;
     if (event.key === "x" || event.key === "Escape") {
       // Perform the desired action when "x" or "esc" is pressed
-      // if (modal === "upload-guidelines")
-      setModal("none");
+      if (modal === "upload-guidelines") {
+        setModal("none");
+      }
+      if (previewUploadedData) setPreviewUploadedData(undefined);
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:159]: I dont want to set modal to none unless its upload-guidelines

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 4]
```diff
@@ -304,7 +333,7 @@ const Upload = ({
             handleValidationErrorResponse(t("ERROR_UPLOADED_FILE"));
           }
           break;
-        case "Shapefile":
+        case SHAPEFILE:
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:336]: But the whole thing is already wrapped in try and catch

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/components/Nagivator.js [hunk 5]
```diff
@@ -49,18 +50,31 @@ const Navigator = (props) => {
...
-        props.nextEventAddon(currentPage, checkDataCompletion);
-      else props.nextEventAddon(currentPage, true);
+    if (checkDataCompletion === "valid" || checkDataCompletion === "invalid") {
+      if (typeof props.nextEventAddon === "function") {
+        if (LoadCustomComponent({ component: props.components[currentPage?.component] }) !== null)
+          props.nextEventAddon(currentPage, checkDataCompletion, setCheckDataCompletion);
+        else props.nextEventAddon(currentPage, true, setCheckDataCompletion);
+      } else {
+        setCheckDataCompletion("perform-action");
+      }
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/components/Nagivator.js:60]: Yes it does

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/components/CommonComponents.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+export const ModalHeading = (props) => {
+  return (
+    <p className={`modal-header ${props.className ? props.className : ""}`} style={props.style}>
+      {props.label}
+    </p>
+  );
+};
+
+export const convertGeojsonToExcelSingleSheet = (InputData, fileName) => {
```
Comment (nipunarora-eGov) [inline micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/components/CommonComponents.js:35]: @siddhant-nawale-egov Put functions in utils

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/utils/index.js [hunk 7]
```diff
@@ -70,11 +71,134 @@ const computeGeojsonWithMappedProperties = ({campaignType,fileType,templateIdent
...
+  }
+
+  // Extract corresponding values for each feature
+  const values = InputData?.map((feature) => {
+    return keys.map((key) => feature.properties[key]);
+  });
+
+  // Group keys and values into the desired format
+  return { [fileName]: [keys, ...values] };
+};
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/utils/index.js:99]: There is no need to console the error here. Its handled some where else.

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Mapping.js [hunk 8]
```diff
@@ -184,11 +215,18 @@ const Mapping = ({
...
+    // setting bounds if they exist
+    let bounds = findBounds(geojsons);
+    if (bounds) {
+      map?.fitBounds(bounds);
+    } else {
+      if (Boundary) {
+        let bounds = findBounds(Boundary);
+        if (bounds) map?.fitBounds(L.latLngBounds(bounds));
+      }
+    }
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Mapping.js:229]: I tested it with files that create error. If the file is not valid it doesnt load and doesnt create run time error

**Review by nipunarora-eGov — APPROVED** [review]

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Mapping.js [hunk 9]
```diff
@@ -265,77 +303,23 @@ const BoundarySelection = memo(
...
-            });
-          }
-        });
-        setProcessedHierarchy(TempHierarchy);
-        setBoundarySelections({ ...oldSelections, ...tempData });
-      },
-      [boundaryData, boundarySelections, hierarchy, processedHierarchy, setBoundarySelections, setProcessedHierarchy]
-    );
+      setIsLoading(false);
+    }, [boundaryData, hierarchy, boundarySelections]);
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Mapping.js:318]: i dont want it to render when processedHierarchy changes

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Hypothesis.js [hunk 10]
```diff
@@ -40,37 +43,74 @@ const Hypothesis = ({ campaignType = "SMC", microplanData, setMicroplanData, che
...
-  }, [assumptions]);
+    let check = assumptions.every((item) => Object.values(item).every((data) => data !== ""));
+    check = check && assumptions.length !== 0;
+    if (check) setCheckDataCompletion("valid");
+    else setCheckDataCompletion("invalid");
+  };
+  const cancelUpdateData = () => {
+    setCheckDataCompletion("false");
+    setModal("none");
+  };
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Hypothesis.js:92]: I am removing it on line number 67

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js [hunk 11]
```diff
@@ -88,19 +93,41 @@ const Upload = ({
...
 
+  // Set TourSteps
+  useEffect(() => {
+    const tourData = tourSteps(t)?.[page] || {};
+    if (state?.tourStateData?.name === page) return;
+    dispatch({
+      type: "SETINITDATA",
+      state: { tourStateData: tourData },
+    });
+  }, [t]);
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Upload.js:104]: I just want this to render on first render

File: micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Mapping.js [hunk 12]
```diff
@@ -81,13 +98,24 @@ const Mapping = ({
...
 
+  // Set TourSteps
+  useEffect(() => {
+    const tourData = tourSteps(t)?.[page] || {};
+    if (state?.tourStateData?.name === page) return;
+    dispatch({
+      type: "SETINITDATA",
+      state: { tourStateData: tourData },
+    });
+  }, []);
```
[the remark this answers is not in the document]
Comment (siddhant-nawale-egov) [reply micro-ui/web/micro-ui-internals/packages/modules/hcm-microplanning/src/pages/employee/Mapping.js:114]: I want it to render on just first render

**Pull request merged**